batch/batch/worker/worker.py

```diff
@@ -2697,6 +2700,38 @@ async def cleanup_old_images(self):
         except Exception as e:
             log.exception(f'Error while deleting unused image: {e}')
 
+    async def send_billing_update(self):
+        async def update():
+            now = time_msecs()
+            billing_update_data = []
+            for (batch_id, job_id), job in self.jobs.items():
+                if job.start_time is None:
+                    continue
+
+                if job.end_time is None:
+                    rollup_time = now
+                else:
+                    rollup_time = job.end_time
+
+                billing_update_data.append(
+                    {
+                        'batch_id': batch_id,
+                        'job_id': job_id,
+                        'attempt_id': job.attempt_id,
+                        'start_time': job.start_time,
+                        'rollup_time': rollup_time,
```

Comment:
We are using the time in the database as the "billing date" because we want to record when the billing charges were collected/accounted for and not when the money was actually spent. This is in contrast to the other ideas we thought about where we split up the time segment into the corresponding dates where the money was spent and using the "rollup time" as the date billed even if part of the spending occurred in a different day. We decided that using the date at which the spending was accounted for has the desired property that past billing charges never fluctuate despite network failures or Batch going offline. It is also simpler to reason about exactly for the reason you cite about retries potentially delaying updates.

Reply:
OK, the distinction between "spend date" and "billing date" helped